library/lua/gui/dwarfmode.lua

```diff
@@ -436,21 +438,61 @@ MenuOverlay = defclass(MenuOverlay, DwarfOverlay)
 MenuOverlay.ATTRS {
     frame_inset = 0,
     frame_background = gui.CLEAR_PEN,
+
+    -- if sidebar_mode is set, we will enter the specified sidebar mode on show
+    -- and restore the previous sidebar mode on dismiss. otherwise it is up to
+    -- the caller to ensure we are in a sidebar mode where the menu is visible.
+    sidebar_mode = DEFAULT_NIL,
 }
 
 function MenuOverlay:computeFrame(parent_rect)
     return self.df_layout.menu, gui.inset_frame(self.df_layout.menu, self.frame_inset)
 end
 
-function MenuOverlay:onAboutToShow(below)
-    MenuOverlay.super.onAboutToShow(self,below)
+function MenuOverlay:onAboutToShow(parent)
+    if not dfhack.isMapLoaded() then
+        qerror('Please load a fortress map.')
+    end
+
+    self.saved_sidebar_mode = df.global.ui.main.mode
+    -- what mode should we restore when this window is dismissed? ideally, we'd
+    -- restore the mode that the user has set, but we should fall back to
+    -- restoring the default mode if either of the following conditions are
+    -- true:
+    -- 1) enterSidebarMode doesn't support getting back into the current mode
+    -- 2) a dfhack viewscreen is currently visible. in this case, we can't trust
+    --    that the current sidebar mode was set by the user. it could just be a
+    --    MenuOverlay subclass that is currently being shown that has set the
+    --    sidebar mode for its own purposes.
+    if not SIDEBAR_MODE_KEYS[self.saved_sidebar_mode]
+            or dfhack.gui.getCurFocus(true):find('^dfhack/') then
+        self.saved_sidebar_mode = df.ui_sidebar_mode.Default
+    end
+    enterSidebarMode(df.ui_sidebar_mode.Default)
+
+    -- refresh parent since the original parent may have just been dismissed
+    parent = dfhack.gui.getCurViewscreen(true)
+    MenuOverlay.super.onAboutToShow(self, parent)
```

Comment:
Your comment above says
```
    -- if sidebar_mode is set, we will enter the specified sidebar mode on show
    -- and restore the previous sidebar mode on dismiss. otherwise it is up to
    -- the caller to ensure we are in a sidebar mode where the menu is visible
```
but this is unconditionally modifying `ui.main.mode`, potentially changing behavior unexpectedly for clients (not 100% sure if it is a "breaking" change). Could this entire block be moved under the check for `self.sidebar_mode`? I think clients could still opt-in by setting `sidebar_mode = Default` where the subclass of MenuOverlay is defined.

Reply:
You're absolutely right. I'll solidify the logic on this before I merge.

Reply:
done